core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp

```diff
@@ -0,0 +1,570 @@
+    // FIXME_SYCL
+    Kokkos::abort("Not implemented!");
+    return 0;
+  }
+
+  template <class ClosureType, class FunctorType>
+  int internal_team_size_max(const FunctorType& /*f*/) const {
+    return m_space.impl_internal_space_instance()->m_maxThreadsPerSM;
+  }
+
+  template <class ClosureType, class FunctorType>
+  int internal_team_size_recommended(const FunctorType& /*f*/) const {
+    // FIXME_SYCL improve
+    return m_space.impl_internal_space_instance()->m_maxThreadsPerSM;
+  }
+};
+
+template <typename FunctorType, typename... Properties>
+class ParallelFor<FunctorType, Kokkos::TeamPolicy<Properties...>,
+                  Kokkos::Experimental::SYCL> {
+ public:
+  using Policy = TeamPolicyInternal<Kokkos::Experimental::SYCL, Properties...>;
+  using functor_type = FunctorType;
+  using size_type    = ::Kokkos::Experimental::SYCL::size_type;
+
+ private:
+  using member_type   = typename Policy::member_type;
+  using work_tag      = typename Policy::work_tag;
+  using launch_bounds = typename Policy::launch_bounds;
+
+  FunctorType const m_functor;
+  Policy const m_policy;
+  size_type const m_league_size;
+  int m_team_size;
+  size_type const m_vector_size;
+  int m_shmem_begin;
+  int m_shmem_size;
+  void* m_scratch_ptr[2];
+  int m_scratch_size[2];
+
+  template <typename Functor>
+  void sycl_direct_launch(const Policy& policy, const Functor& functor) const {
+    // Convenience references
+    const Kokkos::Experimental::SYCL& space = policy.space();
+    Kokkos::Experimental::Impl::SYCLInternal& instance =
+        *space.impl_internal_space_instance();
+    sycl::queue& q = *instance.m_queue;
+
+    q.submit([&](sycl::handler& cgh) {
+      const auto shmem_begin  = m_shmem_begin;
+      const auto shmem_size   = m_shmem_size;
+      const auto scratch_size = m_scratch_size[1];
+      auto team_lambda        = [=](sycl::group<1> g) {
+        // FIXME_SYCL Add scratch memory
+        const member_type team_member(nullptr, shmem_begin, shmem_size, nullptr,
+                                      scratch_size, g);
+
+        if constexpr (std::is_same<work_tag, void>::value)
+          functor(team_member);
+        else
+          functor(work_tag(), team_member);
+      };
+      if (m_team_size > 0)
+        cgh.parallel_for_work_group(sycl::range<1>(m_league_size),
+                                    sycl::range<1>(m_team_size), team_lambda);
+      else
+        cgh.parallel_for_work_group(sycl::range<1>(m_league_size), team_lambda);
+    });
+    space.fence();
+  }
+
+  // Indirectly launch a functor by explicitly creating it in USM shared memory
+  void sycl_indirect_launch() const {
+    // Convenience references
+    const Kokkos::Experimental::SYCL& space = m_policy.space();
+    Kokkos::Experimental::Impl::SYCLInternal& instance =
+        *space.impl_internal_space_instance();
+    Kokkos::Experimental::Impl::SYCLInternal::IndirectKernelMemory& kernelMem =
```

Comment:
Just pointing out this will be another place that will have to be updated with indirect launch changes.

Reply:
Nevin is referring to #3511